synapse/handlers/message.py

```diff
@@ -225,6 +237,76 @@ def get_joined_members(self, requester, room_id):
             for user_id, profile in iteritems(users_with_profile)
         }
 
+    @defer.inlineCallbacks
+    def schedule_deletion_expired(self, event_id, redaction_ts):
+        """Schedule the deletion of an expired event, or if that event should have
+        expired then delete the event immediately.
+
+        Args:
+            event_id (str): The ID of the event to schedule the deletion of.
+            redaction_ts (int): The timestamp to delete the event at.
+        """
+        if not self._ephemeral_events_enabled:
+            return
+
+        # Save the timestamp at which the event expires so that we can reschedule its
+        # deletion on startup if the server is stopped before the event is deleted.
+        yield self.store.insert_event_expiry(event_id, redaction_ts)
+
+        now_ms = self.clock.time_msec()
+        delay = (redaction_ts - now_ms) / 1000
+
+        if delay > 0:
+            # Figure out how many seconds we need to wait before redacting the event.
+            logger.info("Scheduling deletion of event %s in %.3fs", event_id, delay)
+            self.clock.call_later(delay, self._delete_expired_event, event_id)
```

Comment:
I'm slightly concerned that we could accumulate a (very) large number of these callbacks, if we have long expiry times and a large number of events.

Maybe it would be better just to have one at a time, whose scheduled time is that of the earliest expiring event. Then it can check the db table for events which should have expired, expire them all, and reschedule itself?